lib/wallaroo/core/sink/connector_sink/connector_sink.pony

```diff
@@ -301,20 +389,119 @@ actor ConnectorSink is Sink
+      Fail()
+    end
+    if txn_id != _twopc_txn_id then
+      @printf[I32]("2PC: ERROR: twopc_reply: txn_id %s != %s\n".cstring(),
+        txn_id.cstring(), _twopc_txn_id.cstring())
+      Fail()
+    end
+
+    _twopc_phase1_commit = commit
+    if commit then
+      // NOTE: TwoPCFsm2Commit means that our connector sink has
+      // voted to commit, and that it does not mean that we know
+      // the status of the global Wallaroo checkpoint protocol.
+      _twopc_state = cp.TwoPCFsm2Commit
+      ifdef "checkpoint_trace" then
+        @printf[I32]("2PC: txn_id %s was %s\n".cstring(), txn_id.cstring(), commit.string().cstring())
+      end
+
+      _barrier_coordinator.ack_barrier(this, _twopc_barrier_token)
+    else
+      _abort_decision("phase 1 ABORT", _twopc_txn_id, _twopc_barrier_token)
+    end
+
+  fun ref _abort_decision(reason: String, txn_id: String,
+    barrier_token: CheckpointBarrierToken)
+  =>
+    ifdef "checkpoint_trace" then
+      @printf[I32]("2PC: _abort_decision: txn_id %s %s\n".cstring(), txn_id.cstring(), reason.cstring())
+    end
+
+    _twopc_state = cp.TwoPCFsm2Abort
+    _barrier_coordinator.abort_barrier(barrier_token)
+
+  fun _send_phase2(conn: WallarooOutgoingNetworkActor ref,
+    txn_id: String, commit: Bool)
+  =>
+    let b: Array[U8] val = cp.TwoPCEncode.phase2(txn_id, commit)
+    try
+      let msg: cp.MessageMsg = cp.MessageMsg(0, cp.Ephemeral(), 0, 0, None, [b])?
+       _notify.send_msg(conn, msg)
+     else
+       Fail()
+    end
+    ifdef "checkpoint_trace" then
+      @printf[I32]("2PC: sent phase 2 commit=%s for txn_id %s\n".cstring(), commit.string().cstring(), txn_id.cstring())
+    end
+
+  fun ref twopc_intro_done() =>
+    """
+    This callback is used by ConnectorSinkNotify when the 2PC intro
+    part of the connector sink protocol has finished.
+
+    We use _twopc_barrier_token_at_close to determine if we were
+    disconnected during a round of 2PC.  If so, we assume that we
+    lost the phase1 reply from the connector sink, so we make the
+    pessimistic assumption that the connector sink voted rollback/abort.
+    """
+    if _twopc_barrier_token_at_close != _twopc_barrier_token_initial then
+
+      _abort_decision("TCP connection closed during 2PC",
+        _twopc_txn_id_at_close, _twopc_barrier_token_at_close)
+      ifdef "checkpoint_trace" then
+        @printf[I32]("2PC: Wallaroo local abort for txn_id %s barrier %s\n".cstring(), _twopc_txn_id_at_close.cstring(), _twopc_barrier_token_at_close.string().cstring())
+      end
+
+      _reset_2pc_state()
+      _twopc_txn_id_at_close = ""
+      _twopc_barrier_token_at_close = _twopc_barrier_token_initial
+    end
+
   ///////////////
   // BARRIER
   ///////////////
   be receive_barrier(input_id: RoutingId, producer: Producer,
     barrier_token: BarrierToken)
   =>
+    """
+    Process a barrier token of some type: autoscale, checkpoint,
+    rollback, etc.  We will get lots of them for a specific event,
```

Comment:
This is potentially misleading.  For a particular barrier, we should receive exactly one per input (blocking each input over which we receive one), at which point we can unblock all inputs and continue.

Reply:
Updated